linker.go

```diff
@@ -1,57 +1,149 @@
+			if prog == dep {
 				continue
 			}
 
-			needed, err := needSection(insns, lib.Instructions)
+			need, err := needProg(prog.Instructions, dep.Instructions)
 			if err != nil {
-				return fmt.Errorf("linking %s: %w", lib.Name, err)
+				return fmt.Errorf("dependency-checking program '%s' and '%s': %w", prog.Name, dep.Name, err)
 			}
 
-			if !needed {
+			if !need {
 				continue
 			}
 
-			linked[lib] = true
-			prog.Instructions = append(prog.Instructions, lib.Instructions...)
-			pending = append(pending, lib.Instructions)
-
-			if prog.BTF != nil && lib.BTF != nil {
-				if err := prog.BTF.Append(lib.BTF); err != nil {
-					return fmt.Errorf("linking BTF of %s: %w", lib.Name, err)
-				}
-			}
+			// Register a direct dependency.
+			prog.neighbours = append(prog.neighbours, dep)
 		}
 	}
 
 	return nil
 }
 
-func needSection(insns, section asm.Instructions) (bool, error) {
-	// A map of symbols to the libraries which contain them.
-	symbols, err := section.SymbolOffsets()
+// collectInstructions returns the instruction stream of all progs in order.
+func collectInstructions(progs []*ProgramSpec) asm.Instructions {
+	var out asm.Instructions
+
+	for _, prog := range progs {
+		out = append(out, prog.Instructions...)
+	}
+
+	return out
+}
+
+// collectFuncInfos returns the BTF func infos of all progs in order.
+func collectFuncInfos(progs []*ProgramSpec) ([]byte, error) {
+	if len(progs) == 0 {
+		return nil, nil
+	}
+
+	var off uint64
+
+	buf := bytes.NewBuffer(make([]byte, 0, binary.Size(&btf.FuncInfo{})*len(progs)))
+	for _, prog := range progs {
+		if err := prog.BTF.FuncInfo.Marshal(buf, off); err != nil {
+			return nil, fmt.Errorf("marshaling prog %s func info: %w", prog.Name, err)
+		}
+		off += prog.Instructions.Size()
+	}
+
+	return buf.Bytes(), nil
+}
+
+// collectLineInfos returns the BTF line infos of all progs in order.
+func collectLineInfos(progs []*ProgramSpec) ([]byte, error) {
+	if len(progs) == 0 {
+		return nil, nil
+	}
+
+	var off uint64
+
+	buf := bytes.NewBuffer(make([]byte, 0, binary.Size(&btf.LineInfo{})*len(progs)))
+	for _, prog := range progs {
+		if err := prog.BTF.LineInfos.Marshal(buf, off); err != nil {
+			return nil, fmt.Errorf("marshaling prog %s line infos: %w", prog.Name, err)
+		}
+		off += prog.Instructions.Size()
+	}
+
+	return buf.Bytes(), nil
+}
+
+// flatten returns a unique list of programs that need to be
+// included in spec's bytestream when inserting it into the kernel.
+//
+// It recursively collects spec's dependencies by stepping into its neighboring
+// ProgramSpecs. Each visited program is recorded to both avoid infinite
+// recursion and to prevent collecting the same program more than once.
+//
+// Always returns spec itself as the first entry. Returns nil when spec was
+// already visited.
+//
+// TODO: The stability of the order of the output matters,
+// this needs to be clear and tested.
+func (spec *ProgramSpec) flatten(visited map[*ProgramSpec]bool) []*ProgramSpec {
```

Comment:
> why isn't `spec.neighbours` already this unique list?

For the same reason `Instructions` no longer contains all flattened dependencies: because it causes duplicates. 'Uniqueness' depends on the starting point of the dependency resolution process.

If every function has at most one caller, this is not an issue:
```
A --> B --> C
```
In the example above, B would include C and A would include both B and C. If B gets resolved first, it will include C's instructions, which A would pull in. Care needs to be taken that B is fully resolved before it is included by any callers, otherwise it wouldn't contain C yet.

Things get tricky when a function D is added that calls both A and B:
```
D -|-> A --> B --> C
   |-> B --> C
```
B and C's instructions were already pulled in indirectly by pulling in A, and subsequently pulling in B will cause the same insns to be pulled in again.

Now if C would call back into A, this gets even more complicated, since all of A, B and C now need to include each other's instructions. In order to pre-calculate this, we'd need to execute a full dependency resolution pass for each node in the graph, starting with the deepest node(s) first. But if there are cycles, how do we determine the deepest node?

tl;dr - `neighbours` solves this efficiently by pre-calculating adjacency and only performing traversal when the root of the graph is known (e.g. when inserting an entrypoint prog)

Reply:
> This should be in prog.go if it's on ProgramSpec.

Done